src/Messages/Attachments/OWSAttachmentsProcessor.m

```diff
@@ -233,22 +233,64 @@ - (void)downloadFromLocation:(NSString *)location
     manager.responseSerializer = [AFHTTPResponseSerializer serializer];
     manager.completionQueue    = dispatch_get_main_queue();
 
+    // We want to avoid large downloads from a compromised or buggy service.
+    const long kMaxDownloadSize = 150 * 1024 * 1024;
     // TODO stream this download rather than storing the entire blob.
-    [manager GET:location
-      parameters:nil
-        progress:nil // TODO show some progress!
-         success:^(NSURLSessionDataTask *_Nonnull task, id _Nullable responseObject) {
-             if (![responseObject isKindOfClass:[NSData class]]) {
-                 DDLogError(@"%@ Failed retrieval of attachment. Response had unexpected format.", self.tag);
-                 NSError *error = OWSErrorMakeUnableToProcessServerResponseError();
-                 return failureHandler(task, error);
-             }
-             successHandler((NSData *)responseObject);
-         }
-         failure:^(NSURLSessionDataTask *_Nullable task, NSError *_Nonnull error) {
-             DDLogError(@"Failed to retrieve attachment with error: %@", error.description);
-             return failureHandler(task, error);
-         }];
+    __block NSURLSessionDataTask *task = nil;
+    // We only need to check the content length header once.
+    __block BOOL hasCheckedContentLength = NO;
+    task = [manager GET:location
+        parameters:nil
+        progress:^(NSProgress *_Nonnull progress) {
```

Comment:
nit: I'm not sure what kind of guarantees we have about when `progress` gets called. Probably it gets called at least once? But from what I can see, the NSURLSessionDataDelegate is *the way*™; to inspect the headers of a cancelable data task. 

https://developer.apple.com/reference/foundation/nsurlsessiondatadelegate/1410027-urlsession?language=objc

Also it looks like using the NSURLSessionDataDelegate would pave the way to one day graduate to a proper file download task instead of a data task, which I *think* would allow us to safely download larger attachments in the background *and* allow us to pass around streams/file handles instead of large blobs of binary data. 🎉  

WDYT?